src/states/joining_peer.rs

```diff
@@ -0,0 +1,300 @@
+    peer_map::PeerMap,
+    quic_p2p::{NodeInfo, Peer},
+    routing_message_filter::RoutingMessageFilter,
+    routing_table::Authority,
+    state_machine::{State, Transition},
+    time::Instant,
+    timer::Timer,
+    xor_name::XorName,
+    NetworkService,
+};
+use std::{
+    fmt::{self, Display, Formatter},
+    net::SocketAddr,
+    time::Duration,
+};
+
+// Time (in seconds) after which bootstrap is cancelled (and possibly retried).
+const JOIN_TIMEOUT: Duration = Duration::from_secs(120);
+
+// State of a node after bootstrapping, while joining a section
+pub struct JoiningPeer {
+    network_service: NetworkService,
+    routing_msg_filter: RoutingMessageFilter,
+    msg_backlog: Vec<RoutingMessage>,
+    full_id: FullId,
+    min_section_size: usize,
+    peer_map: PeerMap,
+    timer: Timer,
+    join_token: u64,
+}
+
+impl JoiningPeer {
+    pub fn new(
+        network_service: NetworkService,
+        full_id: FullId,
+        min_section_size: usize,
+        timer: Timer,
+        peer_map: PeerMap,
+        node_infos: Vec<NodeInfo>,
+    ) -> Self {
+        let join_token = timer.schedule(JOIN_TIMEOUT);
+
+        let mut joining_peer = Self {
+            network_service,
+            routing_msg_filter: RoutingMessageFilter::new(),
+            msg_backlog: vec![],
+            full_id,
+            min_section_size,
+            timer: timer,
+            peer_map,
+            join_token,
+        };
+
+        for node_info in node_infos {
+            joining_peer.send_join_request(node_info);
+        }
+        joining_peer
+    }
+
+    pub fn into_adult(
+        self,
+        gen_pfx_info: GenesisPfxInfo,
+        outbox: &mut dyn EventBox,
+    ) -> Result<State, RoutingError> {
+        let dev_config = config_handler::get_config().dev.unwrap_or_default();
+        let mut peer_mgr = PeerManager::new(dev_config.disable_client_rate_limiter);
+
+        // initialise known peers to PeerState::Connected in the PeerManager
+        for pub_id in self.peer_map.connected_ids() {
+            peer_mgr.insert_peer(*pub_id, PeerState::Connected);
+        }
+
+        let details = AdultDetails {
+            network_service: self.network_service,
+            event_backlog: vec![],
+            full_id: self.full_id,
+            gen_pfx_info,
+            min_section_size: self.min_section_size,
+            msg_backlog: self.msg_backlog,
+            peer_map: self.peer_map,
+            peer_mgr,
+            routing_msg_filter: self.routing_msg_filter,
+            timer: self.timer,
+        };
+        Adult::from_joining_peer(details, outbox).map(State::Adult)
+    }
+
+    pub fn into_bootstrapping(self) -> Result<State, RoutingError> {
+        Ok(State::BootstrappingPeer(BootstrappingPeer::new(
+            self.network_service,
+            FullId::new(),
+            self.min_section_size,
+            self.timer,
+        )))
+    }
+
+    fn send_join_request(&mut self, dst: NodeInfo) {
```

Comment:
There is `Base::send_direct_message` which looks like it could be used instead of this whole function.

Reply:
Ah - the problem here is that we only have a `NodeInfo` as the destination, and `Base::send_direct_message` requires a `PublicId`. Not sure if it's worth adding another method to `Base` (one that would take a `NodeInfo`)?

Reply:
Ah yes, I see. Then it's fine.